src/ZkFold/Symbolic/Data/UInt.hs

```diff
@@ -127,522 +102,417 @@ cast n =
+            solve :: MonadCircuit i a m => Vector (NumberOfRegisters (BaseField c) n r) i -> m (Vector  (NumberOfRegisters (BaseField c) k r) i)
+            solve xv = do
+                let regs = V.fromVector xv
                 zeros <- replicateA (value @k -! (value @n)) (newAssigned (Haskell.const zero))
-                extended <- fromBits (highRegisterSize @a @k @r) (registerSize @a @k @r) (zeros <> bsBits)
+                bsBits <- toBits (Haskell.reverse regs) (highRegisterSize @(BaseField с) @n @r) (registerSize @(BaseField с) @n @r)
+                extended <- fromBits (highRegisterSize @(BaseField с) @k @r) (registerSize @(BaseField с) @k @r) (zeros <> bsBits)
                 return $ V.unsafeToVector $ Haskell.reverse extended
 
 instance
-    ( Arithmetic a
+    ( Symbolic c
     , KnownNat n
     , KnownNat k
     , KnownRegisterSize r
     , k <= n
-    , from ~ NumberOfRegisters a n r
-    , to ~ NumberOfRegisters a k r
-    ) => Shrink (UInt n r (ArithmeticCircuit a)) (UInt k r (ArithmeticCircuit a)) where
-    shrink (UInt ac) = UInt (circuitF solve)
+    , from ~ NumberOfRegisters (BaseField c) n r
+    , to ~ NumberOfRegisters (BaseField c) k r
+    ) => Shrink (UInt n r c) (UInt k r c) where
+    shrink (UInt ac) =  UInt $ symbolicF ac (V.unsafeToVector . V.fromVector ) solve
         where
-            solve :: forall i m. MonadBlueprint i a m => m (Vector to i)
-            solve = do
-                regs <- V.fromVector <$> runCircuit ac
-                bsBits <- toBits (Haskell.reverse regs) (highRegisterSize @a @n @r) (registerSize @a @n @r)
-                shrinked <- fromBits (highRegisterSize @a @k @r) (registerSize @a @k @r) (drop (value @n -! (value @k)) bsBits)
+            solve :: MonadCircuit i (BaseField c) m => Vector from i -> m (Vector to i)
+            solve xv = do
+                let regs = V.fromVector xv
+                bsBits <- toBits (Haskell.reverse regs) (highRegisterSize @(BaseField c) @n @r) (registerSize @(BaseField c) @n @r)
+                shrinked <- fromBits (highRegisterSize @(BaseField c) @k @r) (registerSize @(BaseField c) @k @r) (drop (value @n -! (value @k)) bsBits)
                 return $ V.unsafeToVector $ Haskell.reverse shrinked
 
 instance
-    ( KnownNat n
+    ( Symbolic c
+    , KnownNat n
     , KnownNat r
     , KnownNat (r - 1)
     , KnownNat (r + r)
     , KnownRegisterSize rs
-    , r ~ NumberOfRegisters (BaseField b) n rs
-    , Symbolic b
-    , NFData (b (Vector r))
-    , Ord (Bool b) (UInt n rs b)
-    , AdditiveGroup (UInt n rs b)
-    , Semiring (UInt n rs b)
-    , MultiplicativeMonoid (UInt n rs b)
-    , FromConstant Natural (UInt n rs b)
-    , BitState ByteString n b
-    , Iso (ByteString n b) (UInt n rs b)
-    , Eq (Bool b) (UInt n rs b)
+    , r ~ NumberOfRegisters (BaseField c) n rs
+    , NFData (c (Vector r))
+    , Ord (Bool c) (UInt n rs c)
```

Comment:
All the constraints on `UInt` except for `Iso` have to be inferred from `Symbolic c`